### PR TITLE
Better schema dependencies

### PR DESCRIPTION
**Pull request by oetherington**
#9219 moved indexes and views into the main schema, but this caused some initialization errors with, for instance, indexes that depended on functions. This happened because we're using a naive ordering for elements in the schema and we basically just hope that everything works (which it did until recently).

This PR makes some large changes to the way we build the schema:

- We can now explicitely declare dependencies on extensions, indexes, views, collections and functions (there's also implicit dependencies, so indexes always depend on their collection or view)
- The render the schema, we build a directed acyclic graph containing all of the elements that the schema needs to contain and we perform a topological sort to ensure that all of an objects dependencies are included before the object itself
- As recommended by Robert, I also switched from `pg-formatter` to `sql-formatter` which speeds up `yarn makemigrations` about 10x, and also removes the dependency on perl

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201302964208280/1207278366519425) by [Unito](https://www.unito.io)

### CODE REVIEW

File: packages/lesswrong/lib/collections/gardencodes/views.ts [hunk 1]
```diff
@@ -68,8 +68,6 @@ GardenCodes.addView("usersPrivateGardenCodes", function (terms) {
   }
 })
 
-ensureIndex(GardenCodes, {code: 1, deleted: 1, userId: 1, });
```
Comment (oetherington) [inline packages/lesswrong/lib/collections/gardencodes/views.ts:71]: Duplicate definition

File: packages/lesswrong/lib/collections/posts/views.ts [hunk 2]
```diff
@@ -595,12 +595,6 @@ Posts.addView("daily", (terms: PostsViewTerms) => ({
     sort: {baseScore: -1}
   }
 }));
-ensureIndex(Posts,
```
Comment (oetherington) [inline packages/lesswrong/lib/collections/posts/views.ts:598]: Duplicate definition

File: packages/lesswrong/lib/collections/tags/views.ts [hunk 3]
```diff
@@ -143,8 +143,6 @@ Tags.addView('coreAndSubforumTags', (terms: TagsViewTerms) => {
     },
   }
 });
-ensureIndex(Tags, {deleted: 1, core:1, name: 1});
```
Comment (oetherington) [inline packages/lesswrong/lib/collections/tags/views.ts:146]: Duplicate definition

File: packages/lesswrong/lib/collections/users/views.ts [hunk 4]
```diff
@@ -56,8 +56,6 @@ ensureIndex(Users, {"services.resume.loginTokens": 1});
 ensureIndex(Users, {email: 1}, {sparse: 1, collation: { locale: 'en', strength: 2 }})
 ensureIndex(Users, {'emails.address': 1}, {sparse: 1, unique: true, collation: { locale: 'en', strength: 2 }}) //TODO: Deprecate or change to use email
 
-ensureIndex(Users, {email: 1})
```
Comment (oetherington) [inline packages/lesswrong/lib/collections/users/views.ts:59]: Duplicate definition

File: packages/lesswrong/lib/collections/users/views.ts [hunk 5]
```diff
@@ -304,7 +302,11 @@ void ensureCustomPgIndex(`
     AND "deleted" IS NOT TRUE
     AND "email" IS NOT NULL
     AND fm_has_verified_email(emails);
-`);
+`, {
+  dependencies: [
+    {type: "function", name: "fm_has_verified_email"}
```
Comment (oetherington) [inline packages/lesswrong/lib/collections/users/views.ts:307]: @b0b3rt This should fix the case you ran into yesterday

File: packages/lesswrong/server/postgresFunctions.ts [hunk 6]
```diff
@@ -1,248 +1,305 @@
 import { allUserProfileFields } from "./userProfileUpdates";
 
+export type PostgresFunction = {
+  source: string,
+  overload?: string,
+  dependencies?: SchemaDependency[],
+}
```
Comment (oetherington) [inline packages/lesswrong/server/postgresFunctions.ts:20]: Changes in this file are mainly just indentation to make room for the new fields

File: packages/lesswrong/server/scripts/makeMigrations.ts [hunk 7]
```diff
@@ -55,6 +56,234 @@ const schemaFileHeaderTemplate = `-- GENERATED FILE
...
 
+declare global {
+  type SchemaDependency =
+    {type: "extension", name: PostgresExtension} |
+    {type: "collection", name: CollectionNameString} |
+    {type: "function", name: string} |
+    {type: "view", name: string};
+}
+
+abstract class Node {
```
Comment (oetherington) [inline packages/lesswrong/server/scripts/makeMigrations.ts:67]: We build a graph of "things" to put in the schema, where each separate "thing" inherits from `Node`

File: packages/lesswrong/server/scripts/makeMigrations.ts [hunk 8]
```diff
@@ -55,6 +56,234 @@ const schemaFileHeaderTemplate = `-- GENERATED FILE
...
+    return this.index.getName();
+  }
+
+  getQuery() {
+    return new CreateIndexQuery(this.table, this.index);
+  }
+}
+
+class CustomIndexNode extends Node {
+  private static nameRegex = /^\s*CREATE\s+(UNIQUE\s+)?INDEX\s+(CONCURRENTLY\s+)?(IF\s+NOT\s+EXISTS\s+)?"?([a-zA-Z0-9_]+)/i;
```
Comment (oetherington) [inline packages/lesswrong/server/scripts/makeMigrations.ts:148]: This is far from fool-proof, but should be easy enough to fix if it breaks in the future

File: packages/lesswrong/server/scripts/makeMigrations.ts [hunk 9]
```diff
@@ -55,6 +56,234 @@ const schemaFileHeaderTemplate = `-- GENERATED FILE
...
+    for (const node of nodes) {
+      this.addNode(node);
+    }
+  }
+
+  getOverallHash() {
+    return md5(Object.values(this.nodes).map((n) => n.getHash()).sort().join());
+  }
+
+  linearize(): Node[] {
```
Comment (oetherington) [inline packages/lesswrong/server/scripts/makeMigrations.ts:259]: This is where we do the topological sort

File: packages/lesswrong/server/databaseSettings.ts [hunk 10]
```diff
@@ -8,12 +8,6 @@ import { ensureIndex } from '../lib/collectionIndexUtils';
 
 const runValidateSettings = false
 
-ensureIndex(DatabaseMetadata, {
```
Comment (b0b3rt) [inline packages/lesswrong/server/databaseSettings.ts:11]: Is this redundant?  Seems unlikely to come up, but would be bad if someone accidentally inserted e.g. a duplicate `serverSettings` or something

File: packages/lesswrong/server/scripts/makeMigrations.ts [hunk 11]
```diff
@@ -55,6 +56,234 @@ const schemaFileHeaderTemplate = `-- GENERATED FILE
...
+    const depthFirstSearch = (nodeName: string) => {
+      const node = this.nodes[nodeName];
+      if (!node) {
+        throw new Error(`Invalid node: "${nodeName}"`);
+      }
+
+      if (!unvisited.has(nodeName)) {
+        return;
+      }
+      unvisited.delete(nodeName);
```
Comment (b0b3rt) [inline packages/lesswrong/server/scripts/makeMigrations.ts:272]: Probably doesn't matter in practice, but does this let through circular dependencies?  Seems like it if ran into one, it'd just do an early return on L270 (the same as if it'd run into a dependency that wasn't circular, just shared with something else that had already been visited).

**Review by b0b3rt — APPROVED** [review]
lgtm, thanks!  (though also happy to look at whatever gets cooked up for concurrency handling if that happens in this PR too)

File: packages/lesswrong/server/databaseSettings.ts (hunk 10 above)
Comment (oetherington) [reply packages/lesswrong/server/databaseSettings.ts:11]: Yeah - there's an identical index defined using `ensureCustomPgIndex` in `collections/databaseMetadata/collection.ts`

**Review by jpaddison3 — APPROVED** [review]
Deferring to Robert's approval, thanks for this, this doesn't seem trivial.

**Comment by jpaddison3** [comment]
> As recommended by Robert, I also switched from pg-formatter to sql-formatter which speeds up yarn makemigrations about 10x, and also removes the dependency on perl

🎉

**Pull request merged**